kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/GCSBatchTableWriter.java

```diff
@@ -0,0 +1,153 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.BigQueryException;
+import com.google.cloud.bigquery.FormatOptions;
+import com.google.cloud.bigquery.Job;
+import com.google.cloud.bigquery.JobInfo;
+import com.google.cloud.bigquery.LoadJobConfiguration;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.gson.Gson;
+
+import org.apache.kafka.connect.errors.ConnectException;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Batch Table Writer that uploads records to GCS as a blob
+ * and then triggers a load job from that GCS file to BigQuery.
+ */
+public class GCSBatchTableWriter implements Runnable {
+  private static final Gson gson = new Gson();
+  private static final int NOT_FOUND_ERROR_CODE = 404;
+
+  private final BlobInfo blobInfo;
+  private final Storage storage;
+  private final String sourceUri;
+
+  private final BigQuery bigQuery;
+  private final LoadJobConfiguration loadJobConfiguration;
+
+  private final List<Map<String, Object>> records;
+
+  /**
+   * Initializes a batch table writer with a full list of records to write.
+   * @param bucketName The name of the bucket to upload the blob in
+   * @param blobName Full path within the bucket to the blob (without the extension)
+   * @param storage GCS Storage
+   * @param tableId {@link TableId} of the BigQuery table to upload to
+   * @param bigQuery {@link BigQuery} Object used to perform upload
+   * @param records Records to upload to BigQuery through GCS
+   */
+  public GCSBatchTableWriter(String bucketName,
+                             String blobName,
+                             Storage storage,
+                             TableId tableId,
+                             BigQuery bigQuery,
+                             List<Map<String, Object>> records) {
+    BlobId blobId = BlobId.of(bucketName, blobName + ".json");
+    blobInfo = BlobInfo.newBuilder(blobId).setContentType("text/json").build();
+    this.storage = storage;
+    this.sourceUri = String.format("gs://%s/%s.json", bucketName, blobName);
+
+    this.bigQuery = bigQuery;
+
+    // Check if the table specified exists
+    try {
+      if (!bigQuery.getTable(tableId).exists()) {
+        throw new BigQueryException(NOT_FOUND_ERROR_CODE, "");
+      }
+    } catch (BigQueryException | NullPointerException exception) {
+      throw new BigQueryException(NOT_FOUND_ERROR_CODE,
+          String.format("Table id for table %s does not exist", tableId.getTable()),
```

Comment:
It's worth noting that within the normal flow of the connector, on startup we will attempt to create tables.

So, it's fine to throw an error here, but maybe a comment saying that this really really shouldn't happen (just so we know that it's really unexpected in case it ever does show up)